File: python/guidos-gorgeous-lasagna/lasagna.py
"""Functions used in preparing Guido's gorgeous lasagna.

Learn about Guido, the creator of the Python language: https://en.wikipedia.org/wiki/Guido_van_Rossum
"""
EXPECTED_BAKE_TIME: int = 40  # Time in minutes
PREPARATION_TIME: int = 2  # Time in minutes / layer


def bake_time_remaining(elapsed_bake_time: int) -> int:
    """Calculate the bake time remaining.

    :param elapsed_bake_time: int - baking time already elapsed.
    :return: int - remaining bake time (in minutes) derived from 'EXPECTED_BAKE_TIME'.

    Function that takes the actual minutes the lasagna has been in the oven as
    an argument and returns how many minutes the lasagna still needs to bake
    based on the `EXPECTED_BAKE_TIME`.
    """
    return EXPECTED_BAKE_TIME - elapsed_bake_time


def preparation_time_in_minutes(number_of_layers: int) -> int:
    """
    Return preparation time in minutes.

    :param number_of_layers: int
    :return: int - preparation time in minutes.

    This function calculates the preparation time in minutes as a function
    of the number of layers and the constant time per layer defined in 'PREPARATION_TIME'.
    """
    return PREPARATION_TIME * number_of_layers


def elapsed_time_in_minutes(number_of_layers: int, elapsed_bake_time: int) -> int:
    """
    Return elapsed cooking time.

    :param number_of_layers: int - number of layers in the lasagna to prepare.
    :param elapsed_bake_time: int - baking time already elapsed.
    :return: int - elapsed time in minutes.

    This function takes two numbers representing the number of layers & the time already spent
    baking and calculates the total elapsed minutes spent cooking the lasagna.
    """
    return preparation_time_in_minutes(number_of_layers) + elapsed_bake_time

File: elixir/lasagna/lib/lasagna.ex
defmodule Lasagna do
  @moduledoc """
  Bake a lasagna!
  """

  @spec expected_minutes_in_oven() :: 40
  def expected_minutes_in_oven, do: 40

  @spec remaining_minutes_in_oven(number()) :: number()
  def remaining_minutes_in_oven(t), do: expected_minutes_in_oven() - t

  @spec preparation_time_in_minutes(number()) :: number()
  def preparation_time_in_minutes(layers), do: layers * 2

  @spec total_time_in_minutes(number(), number()) :: number()
  def total_time_in_minutes(layers, time), do: preparation_time_in_minutes(layers) + time

  @spec alarm() :: String.t()
  def alarm, do: "Ding!"
end
